File: src/kernelprolog/main/SxxMachine/KPTrail.cpp
using namespace std;

#include "KPTrail.h"
#include "../../../machine/SxxMachine/Term.h"

namespace SxxMachine
{

	void KPTrail::push(Term* var)
	{
	  add(var);
	}

	Term* KPTrail::pop()
	{
	  return any_cast<Term*>(remove(size() - 1));
	}

	any KPTrail::remove(const int& i)
	{
	  return stack.erase(stack.begin() + i);
	}

	bool KPTrail::isEmpty()
	{
	  return stack.empty();
	}

	int KPTrail::size()
	{
	  return stack.size();
	}

	void KPTrail::add(Term* x)
	{
	  stack.push_back(x);
	}

//JAVA TO C++ CONVERTER TODO TASK: No base class can be determined:
	KPTrail::KPTrail()
	{
	}

	wstring KPTrail::name()
	{
	  return "trail" + hashCode() % 64;
	}

	wstring KPTrail::pprint()
	{
//JAVA TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to 'toString':
	  return name() + "\n" + __super::toString() + "\n";
	}

	void KPTrail::unwind(const int& to)
	{
	  // IO.mes("unwind TRAIL: "+name()+": "+size()+"=>"+to);
	  // if(to>size())
	  // IO.assertion("unwind attempted from smaller to larger top");
	  for(int i = size() - to; i > 0; i--)
	  {
		Term* O = pop();
		O->undo();
	  }
	}

	wstring KPTrail::stat()
	{
	  return "Trail=" + to_string(size());
	}
}

File: jsrc-kernelprolog/main/SxxMachine/KPTrail.java
package SxxMachine;

import java.util.ArrayList;

/**
 * Implements a stack of undo actions for backtracking, and in particular,
 * resetting a Var's Ref fiels to unbound (i.e. this).
 *
 *
 * @see Undoable
 * @see Var
 */
public class KPTrail extends Trail {

    public KPTrail() {
        super(Prolog.M);
    }

    ArrayList<Undoable> stack = new ArrayList<Undoable>();

    //    public int top() {
    //        int expect = super.top() + 1;
    //        final int size = stack.size();
    //        if (expect != size) {
    //            mismatch();
    //        }
    //        return size;
    //    }

    //    public final void push(Undoable var) {
    //        super.push(var);
    //        stack.add(var);
    //        top();
    //    }
    //
    //    public final Undoable pop() {
    //        final Object remove = stack.remove(stack.size() - 1);
    //        Undoable Other = super.pop();
    //        if (Other != remove) {
    //            mismatch();
    //        }
    //        return (Undoable) remove;
    //    }

    /**
     * Used to undo bindings after unification, if we intend to leave no side
     * effects.
     */

    //    // synchronized
    //    final public void unwind(int to) {
    //        // super.unwind(to);
    //        // IO.mes("unwind TRAIL: "+name()+": "+size()+"=>"+to);
    //        int remove = top() - to;
    //        if (remove < 0)
    //            IO.assertion("unwind attempted from smaller to larger top");
    //        while (remove-- > 0) {
    //            Undoable O = pop();
    //            O.undo();
    //        }
    //    }
    //
    //    private void mismatch() {
    //        Prolog.Break("mismatch ");
    //        // TODO Auto-generated method stub
    //        if (true)
    //            throw new AbstractMethodError("KPTrail.mismatch");
    //    }

    //
    //    public boolean isEmpty() {
    //        assert super.
    //        return stack.isEmpty();
    //    }

    public String name() {
        return "trail" + hashCode() % 64;
    }

    public String pprint() {
        return name() + "\n" + super.toString() + "\n";
    }

    public String stat() {
        return "Trail=" + top();
    }
}
